### PR TITLE
Add generic support for all fluid + barrel types

### PR DESCRIPTION
**Pull request by macintacos**
Added a bunch of fluids that were missing for some mods that I'm using in a new playthrough. Tested locally and it seems to work great.

Notably: I didn't spend any time looking into how to do this in a generic way for any type of modded fluid. I've never really messed with Factorio mods before, so I figured that would be a bit out of my league.

### CODE REVIEW

**Comment by macintacos** [comment]
I'm actually experimenting with a generic way of doing this, because my curiosity got the best of me. If I don't figure something out within a couple of hours, I'll comment here.

**Comment by macintacos** [comment]
I've given up for now 😅 it feels like this is possible though. The trickiest part is actually figuring out what the _first_ technology that unlocks a given fluid. In the details for the various fluid recipe prototypes, it seems like there are multiple `unlock-recipe`'s designated even if they're not the very first tech to unlock it. I would imagine we'd need to build some kind of small "solver" to figure it out, walking up the tech tree until we get to "fluid-handling" (or something like that) and keeping track of which one has the shortest path, and designating that as its true "unlock recipe". I'm doing a bit too much bumbling around in the API right now and I kinda want to get back to playing, so I'll probably pick this back up at another time.

@T3sT3ro Let me know if you'd like to just accept this PR, and maybe revisit the "generic" solution in another PR.

**Comment by macintacos** [comment]
Lmao okay, I shouldn't have jumped the gun - I think I got something working! Will push a PR up and adjust the name of this PR to detail it. Might adjust the version to 2.0, since it is a decent change to functionality. Thoughts on that?

**Comment by T3sT3ro** [comment]
Thank for PR, I'm only looking into it right now because I didn't remember how my dev setup worked and couldn't find the time earlier. I'll try to merge it in today or tomorrow. I have a question though: what does https://mods.factorio.com/mod/BarrelFluidIcons mod do? I tried enabling and disabling it and reading the description several times, but still can't find the difference or see what's changed.

As for the hardcoded technologies vs generic way — I think hardcoded is easier to implement in this case, but if you have a generic way to walk up the graph which wouldn't require explicitly adding compatibility for each mod it would be very nice and I would be happy to merge it :) (not sure how much time I'll have though in the following days so sorry if it won't be immediate ;)

**Comment by macintacos** [comment]
Oh that mod simply adds barrel recipes for _all_ fluids including modded fluids, if it can find them. I started a new playthrough with a ton of new planets, and many of them add their own fluids, so without this mod's functionality I was always presented with a list of like 30 fluid barrels that I hadn't even unlocked yet 😅 

I just pushed up the generic implementation! It seems to work perfectly great on my machine, but it's entirely possible that I missed something. Please let me know if you have any questions. No worries about the timeline - I have it manually installed locally anyway, so I'm fine with waiting lol.

File: src/data-final-fixes.lua [hunk 1]
```diff
@@ -1,49 +1,105 @@
...
+      if type(k) ~= "number" then
+        k = '"' .. k .. '"'
+      end
+      s = s .. "[" .. k .. "] = " .. dumpTable(v) .. ","
+    end
+    return s .. "} "
+  else
+    return tostring(tableToDump)
+  end
+end
```
Comment (T3sT3ro) [inline src/data-final-fixes.lua]: I reckon something like that exists: https://lua-api.factorio.com/latest/classes/LuaHelpers.html#table_to_json

**Comment by T3sT3ro** [comment]
Something is not right:
- water [un]barreling recipes are unlocked by "space platform thruster" instead of "fluid handling"
- crude oil [un]barreling cannot be unlocked anymore (should be unlocked with "oil gathering")

More issues probably occur, I did not check further for now.

**Comment by T3sT3ro** [comment]
Maybe the assumption that all fluids are products of recipes doesn't cover all edge cases like with the crude oil which is mined by pumpjacks. For that case I am leaning towards a hardcoded solution where certain recipes require manual assignment, which is more straightforward and explicit. Generic one requires more testing and consideration for edge cases.

**Comment by macintacos** [comment]
Ah - up to you @T3sT3ro. I just pushed a fix that handles some edge cases like water and crude oil. I manually went through all of the fluids that had barrel recipes and added some manual overrides, and explicitly skipped processing water since there isn't even really a tech associated with it from what I can tell.

Everything seems to be working good now, but let me know if you'd like me to just switch back to the hard-coded approach. I might keep this fork for my own purposes and keep iterating if that's the case 😅

**Review by T3sT3ro — APPROVED** [review]
Great work! Tested and working, I did some final refactor and simplification. Also I remembered this: https://lua-api.factorio.com/latest/auxiliary/libraries.html
